File: code/java/ch20/practice/Warmup02NumberOfIslands.java
package ch20.practice;

import java.util.*;

/**
 * Warmup 2: Number of Islands
 * Chapter 20: Graphs II — Real Problems
 *
 * PROBLEM: Given an m x n grid of 0s and 1s, count the number of islands
 *          (connected components of 1s, 4-directional).
 *
 * INSTRUCTIONS: Replace the body of solve() with your solution.
 */
public class Warmup02NumberOfIslands {
    public static int solve(int[][] grid) {
        // TODO: Replace this with your solution
        return 0;
    }

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        int rows = scan.nextInt(), cols = scan.nextInt();
        int[][] grid = new int[rows][cols];
        for (int i = 0; i < rows; i++)
            for (int j = 0; j < cols; j++)
                grid[i][j] = scan.nextInt();
        System.out.println(solve(grid));
        scan.close();
    }
}

File: code/python/ch20/practice/warmup_02_number_of_islands.py
"""
Warmup 2: Number of Islands
============================================
Chapter 20: Graphs II -- Real Problems

PROBLEM
-------
Given an m x n grid of 0s (water) and 1s (land), count the number of
islands. An island is a group of 1s connected 4-directionally.

EXAMPLES
--------
>>> solve([[1,1,0,0,0],[1,1,0,0,0],[0,0,1,0,0],[0,0,0,1,1]])
3

CONSTRAINTS
-----------
- 1 <= m, n <= 300
- grid[i][j] is 0 or 1
"""


def solve(grid: list[list[int]]) -> int:
    """Return the number of islands in the grid."""
    pass


# -- Do not change anything below this line ----------------------------------
if __name__ == "__main__":
    first_line = input().split()
    rows, cols = int(first_line[0]), int(first_line[1])
    grid = []
    for _ in range(rows):
        grid.append(list(map(int, input().split())))
    print(solve(grid))
